core/bluetooth/src/main/java/edu/stanford/spezikt/core/bluetooth/domain/BLEServiceImpl.kt

```diff
@@ -0,0 +1,173 @@
+package edu.stanford.spezikt.core.bluetooth.domain
+
+import android.Manifest
+import android.bluetooth.BluetoothAdapter
+import android.bluetooth.BluetoothDevice
+import edu.stanford.spezi.logging.speziLogger
+import edu.stanford.spezi.utils.extensions.append
+import edu.stanford.spezikt.core.bluetooth.api.BLEService
+import edu.stanford.spezikt.core.bluetooth.data.model.BLEDeviceSession
+import edu.stanford.spezikt.core.bluetooth.data.model.BLEServiceEvent
+import edu.stanford.spezikt.core.bluetooth.data.model.BLEServiceState
+import edu.stanford.spezikt.core.bluetooth.data.model.Measurement
+import edu.stanford.spezikt.coroutines.di.Dispatching
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asSharedFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import java.util.concurrent.ConcurrentHashMap
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+internal class BLEServiceImpl @Inject constructor(
+    private val bluetoothAdapter: BluetoothAdapter,
+    private val permissionHandler: PermissionHandler,
+    private val deviceScanner: BLEDeviceScanner,
+    @Dispatching.IO private val scope: CoroutineScope,
+    private val deviceConnectorFactory: BLEDeviceConnector.Factory,
+) : BLEService {
+
+    private val logger by speziLogger()
+
+    private val connectedDevices = ConcurrentHashMap<String, BLEDeviceConnector>()
+    private val _state = MutableStateFlow<BLEServiceState>(BLEServiceState.Idle)
+    private val _events = MutableSharedFlow<BLEServiceEvent>(replay = 1, extraBufferCapacity = 1)
+
+    override val state: StateFlow<BLEServiceState> = _state.asStateFlow()
+    override val events: Flow<BLEServiceEvent> = _events.asSharedFlow()
+
+    override fun start() {
+        logger.i { "start() requested" }
+        when {
+            deviceScanner.isScanning -> logger.i { "Already scanning. Ignoring start request" }
+            bluetoothAdapter.isEnabled.not() -> logger.i { "Start ignored, bluetooth not enabled" }
+            else -> {
+                val missingPermissions = REQUIRED_PERMISSIONS.filterNot { permissionHandler.isPermissionGranted(it) }
+                if (missingPermissions.isNotEmpty()) {
+                    emitEvent(event = BLEServiceEvent.MissingPermissions(permissions = missingPermissions))
+                } else {
+                    startScannerEventCollection()
```

Comment:
Yep, valid point. Before it was starting because on permission change of an app it's activity gets recreated, and we were invoking `start` `onCreate`. I know moved the start in the view model `init` block which only gets created and called once. Both options aren't ideal. The proper way would be reacting on permission changes to invoke the start again as done in the `MyComposable` in this [blog post](https://medium.com/@rzmeneghelo/how-to-request-permissions-in-jetpack-compose-a-step-by-step-guide-7ce4b7782bd7). I will adapt that 💪